### ISSUE TITLE
NoMethodError are reraised as ActionController::RoutingError and result in 404 

### ISSUE THREAD

**Issue by maximeg** [rails/rails#22368]
To reproduce, in any controller method, drop a
```ruby
{}.i_dont_exist
```

The stacktrace points me to:
https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/routing/route_set.rb#L35

As:
```ruby
>> NoMethodError.ancestors
=> [NoMethodError, NameError, StandardError, ...]
```

The refactoring done at 4276b214 may be the cause.

More generally, any `NameError` or its descendants should not be caught here. I think the goal is to only catch `NameError` generated by non existant controller class or `NoMethodError` generated by non existant action method and turn them into `ActionController::RoutingError`.

I'll be happy to dig into this and propose a PR that:
 - raise `ActionController::RoutingError` just around the constantize retrieving the controller class ;
 - raise `ActionController::RoutingError`if that controller class doesn't respond publicly to the requested action. 

I'll just need an enlightenment about `@raise_on_name_error`.

This issue is master only (5.0.0.alpha).

**Comment by jgrau** [comment]
I am having this issue as well.

**Comment by maximeg** [comment]
> I'll be happy to dig into this and propose a PR that:
> - raise `ActionController::RoutingError` just around the constantize retrieving the controller class ;
> - raise `ActionController::RoutingError`if that controller class doesn't respond publicly to the requested action. 

The handling of missing action was not needed because it is already fine, done by `AbstractController::ActionNotFound` error.

I just had to move the rescue of the `NameError` in the dedicated method to rescue more precisely.
I added a test case to ensure that.

**Comment by maximeg** [comment]
Fix in #22564 is merged. Thanks @sgrif :smile:
